File: exercise/kotlin/day11/src/main/kotlin/roman/numerals/RomanNumerals.kt
package roman.numerals

import arrow.core.None
import arrow.core.Option
import arrow.core.Some

object RomanNumerals {
    private const val MAX_NUMBER = 3999
    private fun romanMapping() = mapOf(
        1000 to "M",
        900 to "CM",
        500 to "D",
        400 to "CD",
        100 to "C",
        90 to "XC",
        50 to "L",
        40 to "XL",
        10 to "X",
        9 to "IX",
        5 to "V",
        4 to "IV",
        1 to "I"
    )

    fun convert(number: Int): Option<String> = if (isInRange(number)) Some(convertSafely(number)) else None
    private fun isInRange(number: Int): Boolean = number in 1..MAX_NUMBER
    private fun convertSafely(number: Int): String = romanMapping().let {
        var roman = ""
        var remaining = number

        it.forEach { (key, value) ->
            while (remaining >= key) {
                roman += value
                remaining -= key
            }
        }
        return roman
    }
}

File: exercise/ts/day11/src/romanNumerals.ts
import {none, Option, some} from "fp-ts/Option";

const intToNumerals: Map<number, string> = new Map([
    [1000, "M"],
    [900, "CM"],
    [500, "D"],
    [400, "CD"],
    [100, "C"],
    [90, "XC"],
    [50, "L"],
    [40, "XL"],
    [10, "X"],
    [9, "IX"],
    [5, "V"],
    [4, "IV"],
    [1, "I"],
]);

const maxNumber = 3999;
const isInRange = (number: number): boolean => number > 0 && number <= maxNumber;

export function convert(number: number): Option<string> {
    if (!isInRange(number)) return none;

    let roman = '';
    let remaining = number;

    for (const [value, numeral] of intToNumerals) {
        while (remaining >= value) {
            roman += numeral;
            remaining -= value;
        }
    }
    return some(roman);
}

